### PR TITLE
[KYUUBI #3226] Privileges should be checked only once in `RuleAuthorization`

### PR DESCRIPTION
**Pull request by yikf**
### _Why are the changes needed?_

fix https://github.com/apache/incubator-kyuubi/issues/3226

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Review by ulysses-you — APPROVED** [review]
cc @yaooqinn

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala [hunk 1]
```diff
@@ -22,20 +22,30 @@ import scala.collection.mutable.ArrayBuffer
...
 import org.apache.kyuubi.plugin.spark.authz.{ObjectType, _}
 import org.apache.kyuubi.plugin.spark.authz.ObjectType._
+import org.apache.kyuubi.plugin.spark.authz.ranger.RuleAuthorization.KYUUBI_AUTH_TAG
 import org.apache.kyuubi.plugin.spark.authz.util.AuthZUtils._
 
 class RuleAuthorization(spark: SparkSession) extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = {
-    RuleAuthorization.checkPrivileges(spark, plan)
+    plan match {
+      case p if !p.getTagValue(KYUUBI_AUTH_TAG).getOrElse(false) =>
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala:35]: what if this plan is cached？ for group mode, multiple client users may share the same cache manager, can we add a unit test to verify

Comment (yikf) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala:35]: good catch, will add test suites to verify this.

**Comment by yikf** [comment]
@yaooqinn tests is added to verify the multi-session & cache plan scenario;

Even if plan is cached, check is performed for every new query, this is because `QueryExecution.optimizedPlan` did not copy PlanTag when `withCachedData.clone()` [1], but it seems to be in line with our expectations

[1] [Spark QueryExecution base code](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala#L126)

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 2]
```diff
@@ -59,6 +61,63 @@ abstract class RangerSparkExtensionSuite extends AnyFunSuite
     s"Permission denied: user [$user] does not have [$privilege] privilege on [$resource]"
   }
 
+  test("RuleAuthorization: Should check privileges once only.") {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:64]: let's add `[KYUUBI #3226]` here

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala [hunk 3]
```diff
@@ -59,6 +61,63 @@ abstract class RangerSparkExtensionSuite extends AnyFunSuite
...
+    val rule = new RuleAuthorization(spark)
+
+    (1 until 10).foreach { ignore =>
+      rule.apply(logicalPlan)
+    }
+
+    assert(logicalPlan.getTagValue(KYUUBI_AUTH_TAG).getOrElse(0) == 1)
+  }
+
+  test("RuleAuthorization: Another session should also check even if the plan is cached.") {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RangerSparkExtensionSuite.scala:75]: ditto

**Review by yaooqinn — COMMENTED** [review]
[KYUUBI #3226] needs to be added to title

**Comment by yikf** [comment]
thanks @yaooqinn @ulysses-you  for review, addressed comment

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
please fix code style

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**